File: src/GameLogic.cpp
#include "GameLogic.hpp"

#include <iostream>
#include "engine/entity-system/include.hpp"
#include "engine/game-loop/SingleThreadGameLoop.hpp"
#include "engine/resource-system/include.hpp"
#include "init/load-input-tracker.hpp"
#include "init/load-resources.hpp"
#include "init/register-states.hpp"
#include "ResourceFiles.hpp"
#include "Settings.hpp"

GameLogic::GameLogic(ComponentManager& manager, ResourceStorage& storage)
 : componentManager(manager),
   resourceStorage(storage),
   inputTracker(loadInputTracker(ResourceFiles::CONTROLS)),
   inputDispatcher(inputTracker) {
    gameData = {
        &componentManager,
        &inputDispatcher,
        &resourceStorage,
        &stateMachine,
        nullptr
    };

    Settings& settings = resourceStorage.get<Settings>("settings");

    loadResources(resourceStorage);
    registerStates(gameData);
    stateMachine.pushState(settings.getInitialState());
}

void GameLogic::operator()(SingleThreadGameLoop&, double timeSinceLastFrame) {
    gameData.timeSinceLastFrame = &timeSinceLastFrame;
    inputDispatcher.tick();
    stateMachine.execute();
}

File: frontend/src/GameLogic.ts
import controlsMap from '../../resources/json/controls.json';
import { CoreStructures } from './CoreStructures';
import { GameLoop, UpdateFunction } from './engine/GameLoop';
import { InputDispatcher } from './engine/input-system/InputDispatcher';
import { ResourceStorage } from './engine/resource-system/ResourceStorage';
import { StateMachine } from './engine/state-system/StateMachine';
import { loadInputTracker } from './init/load-input-tracker';

export class GameLogic implements UpdateFunction {
    constructor(resourceStorage: ResourceStorage) {
        const inputTracker = loadInputTracker(controlsMap);
        const inputDispatcher = new InputDispatcher(inputTracker);
        const stateMachine = new StateMachine();

        this.gameData = {
            resourceStorage,
            inputDispatcher,
            stateMachine,
            timeSinceLastFrame: 0
        };
    }

    tick(_: GameLoop, timeSinceLastFrame: number): void {
        this.gameData.timeSinceLastFrame = timeSinceLastFrame;
        this.gameData.inputDispatcher.tick();
        this.gameData.stateMachine.execute();
    }

    private gameData: CoreStructures;
}
